src/inc/internal/AppxBundleWriter.hpp

```
@@ -0,0 +1,90 @@
+
+namespace MSIX {
+    class AppxBundleWriter final : public ComClass<AppxBundleWriter, IBundleWriter, IAppxBundleWriter, IAppxBundleWriter4>
+    {
+    public:
+        AppxBundleWriter(IMsixFactory* factory, const ComPtr<IZipWriter>& zip);
+        ~AppxBundleWriter() {};
+
+        // IBundleWriter
+        void ProcessBundlePayload(const ComPtr<IDirectoryObject>& from, bool flatBundle) override;
+
+        // IAppxBundleWriter
+        HRESULT STDMETHODCALLTYPE AddPayloadPackage(LPCWSTR fileName, IStream* packageStream) noexcept override;
+        HRESULT STDMETHODCALLTYPE Close() noexcept override;
+
+        // IAppxBundleWriter4
+        HRESULT STDMETHODCALLTYPE AddPackageReference(LPCWSTR fileName, IStream* inputStream, 
+            BOOL isDefaultApplicablePackage) noexcept override;
+        HRESULT STDMETHODCALLTYPE AddPayloadPackage(LPCWSTR fileName, IStream* packageStream, 
+            BOOL isDefaultApplicablePackage) noexcept override;
+        HRESULT STDMETHODCALLTYPE AddExternalPackageReference(LPCWSTR fileName, IStream* inputStream,
+            BOOL isDefaultApplicablePackage) noexcept override;
+
+    protected:
+        typedef enum
+        {
+            Open = 1,
+            Closed = 2,
+            Failed = 3
+        }
+        WriterState;
+
+        void ValidateAndAddPayloadFile(const std::string& name, IStream* stream,
+            APPX_COMPRESSION_OPTION compressionOpt, const char* contentType);
+
+        void AddFileToPackage(const std::string& name, IStream* stream, bool toCompress,
+            bool addToBlockMap, const char* contentType, bool forceContentTypeOverride = false);
+
+        void ValidateCompressionOption(APPX_COMPRESSION_OPTION compressionOpt);
+
+        void CloseInternal();
+
+        HRESULT AddPackageReferenceInternal(_In_ LPCWSTR fileName, _In_ IStream* packageStream,
+            _In_ bool isDefaultApplicablePackage);

```

Usually for internal methods we do more of a c++ approach than the com one. For example this method instead of returning a failed HR it will throw an exception that will be caught later on.

---

Updated to return void for most internal methods and use exceptions except for methods implementing an interface which returns HRESULT.